### PR TITLE
[SPARK-30601][BUILD] Add a Google Maven Central as a primary repository

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to address four things. Three issues and fixes were a bit mixed so this PR sorts it out. See also http://apache-spark-developers-list.1001551.n3.nabble.com/Adding-Maven-Central-mirror-from-Google-to-the-build-td28728.html for the discussion in the mailing list.

1. Add the Google Maven Central mirror (GCS) as a primary repository. This will not only help development more stable but also in order to make Github Actions build (where it is always required to download jars) stable. In case of Jenkins PR builder, it wouldn't be affected too much as it uses the pre-downloaded jars under `.m2`.

    - Google Maven Central seems stable for heavy workload but not synced very quickly (e.g., new release is missing)
    - Maven Central (default) seems less stable but synced quickly.

    We already added this GCS mirror as a default additional remote repository at SPARK-29175. So I don't see an issue to add it as a repo.
    https://github.com/apache/spark/blob/abf759a91e01497586b8bb6b7a314dd28fd6cff1/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L2111-L2118

2. Currently, we have the hard-corded repository in [`sbt-pom-reader`](https://github.com/JoshRosen/sbt-pom-reader/blob/v1.0.0-spark/src/main/scala/com/typesafe/sbt/pom/MavenPomResolver.scala#L32) and this seems overwriting Maven's existing resolver by the same ID `central` with `http://` when initially the pom file is ported into SBT instance. This uses `http://` which latently Maven Central disallowed (see https://github.com/apache/spark/pull/27242)

    My speculation is that we just need to be able to load plugin and let it convert POM to SBT instance with another fallback repo. After that, it _seems_ using `central` with `https` properly. See also https://github.com/apache/spark/pull/27307#issuecomment-576720395.

    I double checked that we use `https` properly from the SBT build as well:

    ```
    [debug] downloading https://repo1.maven.org/maven2/com/etsy/sbt-checkstyle-plugin_2.10_0.13/3.1.1/sbt-checkstyle-plugin-3.1.1.pom ...
    [debug] 	public: downloading https://repo1.maven.org/maven2/com/etsy/sbt-checkstyle-plugin_2.10_0.13/3.1.1/sbt-checkstyle-plugin-3.1.1.pom
    [debug] 	public: downloading https://repo1.maven.org/maven2/com/etsy/sbt-checkstyle-plugin_2.10_0.13/3.1.1/sbt-checkstyle-plugin-3.1.1.pom.sha1
    ```

    This was fixed by adding the same repo (https://github.com/apache/spark/pull/27281), `central_without_mirror`, which is a bit awkward. Instead, this PR adds GCS as a main repo, and community Maven central as a fallback repo. So, presumably the community Maven central repo is used when the plugin is loaded as a fallback.

3. While I am here, I fix another issue. Github Action at https://github.com/apache/spark/pull/27279 is being failed. The reason seems to be scalafmt 1.0.3 is in Maven central but not in GCS.

    ```
    org.apache.maven.plugin.PluginResolutionException: Plugin org.antipathy:mvn-scalafmt_2.12:1.0.3 or one of its dependencies could not be resolved: Could not find artifact org.antipathy:mvn-scalafmt_2.12:jar:1.0.3 in google-maven-central (https://maven-central.storage-download.googleapis.com/repos/central/data/)
        at org.apache.maven.plugin.internal.DefaultPluginDependenciesResolver.resolve     (DefaultPluginDependenciesResolver.java:131)
    ``` 

   `mvn-scalafmt` exists in Maven central:

    ```bash
    $ curl https://repo.maven.apache.org/maven2/org/antipathy/mvn-scalafmt_2.12/1.0.3/mvn-scalafmt_2.12-1.0.3.pom
    ```

    ```xml
    <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
        <modelVersion>4.0.0</modelVersion>
        ...
    ```

    whereas not in GCS mirror:

    ```bash
    $ curl https://maven-central.storage-download.googleapis.com/repos/central/data/org/antipathy/mvn-scalafmt_2.12/1.0.3/mvn-scalafmt_2.12-1.0.3.pom
    ```
    ```xml
    <?xml version='1.0' encoding='UTF-8'?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Details>No such object: maven-central/repos/central/data/org/antipathy/mvn-scalafmt_2.12/1.0.3/mvn-scalafmt_2.12-1.0.3.pom</Details></Error>%
    ```

    In this PR, simply make both repos accessible by adding to `pluginRepositories`.

4. Remove the workarounds in Github Actions to switch mirrors because now we have same repos in the same order (Google Maven Central first, and Maven Central second)

### Why are the changes needed?

To make the build and Github Action more stable.

### Does this PR introduce any user-facing change?

No, dev only change.

### How was this patch tested?

I roughly checked local and PR against my fork (https://github.com/HyukjinKwon/spark/pull/2 and https://github.com/HyukjinKwon/spark/pull/3).

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @srowen, @dongjoon-hyun, @tgravescs, @koeninger

**Comment by HyukjinKwon** [comment]
Just for clarification, 1. and 2. in the PR description are already fixed in the master. This PR addresses Sean's comment. Plus, it fixes the 3. in the PR description.

File: pom.xml
Comment (srowen) [inline pom.xml]: Yeah, this is more like what I was thinking. But does it resolve the hard-coded SBT issue? If so, OK.

**Comment by tgravescs** [comment]
for the sbt issue, you are saying if it fails from the first maven repo then it falls back to the default version hardcoded in sbt-pom-reader that has http:?  So if you have another repo in there it first falls back to try that one.  I assume if that one fails then it falls back to the http: one again?

It seems like we should also update the sbt-pom-reader.

**Comment by HyukjinKwon** [comment]
@tgravescs, just for clarification, this change is needed to address all of three. So, even if `sbt-pom-reader` fixes the issue, this change here will still be needed to address other two bullet points in the PR description.

For  the sbt issue, I am a bit hesitant to explain because I am super confident of my answer so just take it as a speculation. _I think_:

1. `sbt-pom-reader` is loaded as a SBT plugin.
2. It tries to convert POM file to SBT project structure
3. During this process, it overwrites the default `central` maven resolver at [here](https://github.com/JoshRosen/sbt-pom-reader/blob/v1.0.0-spark/src/main/scala/com/typesafe/sbt/pom/MavenPomResolver.scala#L32), which is `http`.
4. The conversion is done.
5. Now, we use SBT to resolve the dependencies by the resolvers defined [here](https://github.com/apache/spark/blob/master/project/SparkBuild.scala#L227), which is [`https://... `](https://www.scala-sbt.org/0.13/docs/Resolvers.html).

Before this fix, at 3., `central` is `http` and there's no other repos to fall back. So it fails.
After this fix, at 3., `google-manve-central` becomes a fall back so it can handle during 3.

And yes, currently, I think it's an issue in `sbt-pom-reader` too.

File: pom.xml
Comment (HyukjinKwon) [reply pom.xml]: Yes, just for clarification, https://github.com/apache/spark/commit/c992716a33d78ce0a9aa78b26f5bdb45c26c2a01 fixed both hard-coded SBT issue and Github Action issue (SPARK-30572, SPARK-30534) given my tests. This is just a different way to fix both.

**Comment by srowen** [comment]
Oh yeah good point. Even if it fails to access central over http:// , it should try the Google repo then, this way. That would explain the fix I think.

**Comment by tgravescs** [comment]
Ok, thanks for the explanation, I was just trying to understand how it fixed the SBT issue and what you say makes sense.  

can we file a followup jira to try to fix the sbt-pom-reader as well, that way we don't have to wait for it to fail and fall back.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
In practical, this is a less robust approach than the existing one in the following reasons.
1. The failure rate on Maven Central is much higher than `Google Maven Central` as we experienced in GitHub Action environment until now. So, in terms of the number of fallback call, `Google Maven Central` will be used more frequently as a fallback than the number of fallback call on `Maven Central` in the existing approach.
2. More importantly, if `Maven Central` fails at new dependency (like Chill / Lz4) due to its flakiness, `Google Maven Central` can not provide a fallback because it doesn't have. It will end up a failure. So, the `subset (Google Maven Central)` cannot be a complete fallback of the `superset (Maven Central)` in theory.

   > This is used as a fallback when `central` fails.

I'm -1 for this approach because this is a regression in GitHub Action environment.

**Comment by srowen** [comment]
Hm, but what's the problem with the fallback -- does it take a long time to time out? If it's rare, does it matter? I haven't seen Maven Central being particularly problematic (of course the http:// thing happens 100% of the time in sbt-pom-reader)

If Maven Central doesn't have an artifact, then no neither will the mirror. But, then the artifact effectively doesn't exist. This isn't a problem we can solve, but also isn't the problem here. That is, for example. chill 0.9.5 has always been on Maven Central since release.

The status quo doesn't have any fallback to Google, so I'm not seeing how that can be more robust?

**Comment by dongjoon-hyun** [comment]
@srowen `Maven Central` fails during transportation due to heavy connections. I know it well because I've introduced GitHub Action here and have been monitoring it. It's not a problem of the absence of the artifact at all. `Maven Central` is the ground truth.

You are not considering my second case (2). `Maven Central` can fail on any artifacts like `Chill 0.9.5`. and `Google Maven Central` is not a valid fallback for that as we observed in your PR. Why `subset` can be a fallback for `superset`.

**Comment by srowen** [comment]
I'm confused, are you arguing for 'no fallback'? that can't possibly be better. As you say here: what do you do if you can't reach Maven Central?

The current implementation is basically: try `central` but one that is overridden to point to Google by `<mirror>`, and then try Maven Central under an alias second.

This PR does the same thing, just explicitly. I don't see the difference other than it being simpler?

If chill 0.9.5 were not on Central, it would effectively not exist. Having a mirror doesn't change that, yes. But it is also not the situation here. It is on Central. The Google repo is not here to help this situation. It's here to help your first point.

Conversely, accessing Maven Central is also here to address your second point. You need both at some level to solve both, right?

**Comment by dongjoon-hyun** [comment]
The following is wrong, isn't it?
> This PR does the same thing, just explicitly. I don't see the difference other than it being simpler?

In computer architecture, we try `cache` before `memory`. This PR try `memory` before `cache`.

**Comment by srowen** [comment]
Ah OK right, this PR tries Central, then Google. Right now, the status quo is trying Google then Central. But we can flip the order here. It would be simpler. Would that be OK?

**Comment by dongjoon-hyun** [comment]
Yes. I'm completely okay for that approach `Google Maven Central -> Maven Central`.
So, that was my question to you previously. Did you remember that I was worrying that the community is okay with `Google Maven Central` as a main repo. :)

**Comment by srowen** [comment]
Yes I don't see a problem with it. It's just a mirror, and already being used in the build in one instance.
What is the potential problem you see?

**Comment by dongjoon-hyun** [comment]
It would be great if we can ask it officially in the dev mailing list if @HyukjinKwon and @srowen want to set it a primary repo. I'll give +1 there.

**Comment by srowen** [comment]
I sent an email, but you might want to follow up with what the potential issue is. Others may also not be thinking of what you are.

**Comment by HyukjinKwon** [comment]
Okay, so the main argument to switch the order is:
- `Maven Central` first `Google Maven Central` second is slow because the fallback would happen too much in Github Actions.
- `Google Maven Central` first `Maven Central` second is first.

I don't mind switching the order either; however, I think I have another option:

Why don't we just give a shot to switch the mirrors in `.github/workflows/master.yml`? Not sure if it works but I can test.

```xml
<mirrors>
  <mirror>
   <id>another-central</id>
   <name>Maven Central</name>
   <url>https://repo.maven.apache.org/maven2</url>
   <mirrorOf>central</mirrorOf>
  </mirror> 
  <mirror>
   <id>another-google-maven-central</id>
   <name>Google Maven Central</name>
   <url>https://maven-central.storage.googleapis.com/repos/central/data/</url>
   <mirrorOf>google-maven-central</mirrorOf>
  </mirror> 
</mirrors> 
```

I don't like to list up the same repos twice which look a bit odd. For switching order, I think it's fine but in the way above, we can do it a bit more conservatively.

This fallback issue seems specific to Github Actions and seems best to fix it for Github Actions specifically.

**Comment by HyukjinKwon** [comment]
Alright, seems we're fine to switch. Let me just switch for now. We can consider https://github.com/apache/spark/pull/27307#issuecomment-576951473 as an option later if it becomes problematic for any reason.

**Comment by HyukjinKwon** [comment]
> can we file a followup jira to try to fix the sbt-pom-reader as well, that way we don't have to wait for it to fail and fall back.

@tgravescs, it's actually tricky too because we use our own fork for sbt-pom-reader as of https://github.com/apache/spark/pull/13564. I am not sure if the upstream has the issue or not. We should first revert SPARK-15827, actually. cc @JoshRosen

**Comment by HyukjinKwon** [comment]
> can we file a followup jira to try to fix the sbt-pom-reader as well, that way we don't have to wait for it to fail and fall back.

@tgravescs, it's actually tricky too because we use our own fork for sbt-pom-reader. Actually, we should find a way to don't rely on the fork first. I don't know if the upstream has the issue or not. cc @JoshRosen.

**Comment by JoshRosen** [comment]
@HyukjinKwon,

Here's a code comment explaining the original rationale for the custom POM reader fork: https://github.com/apache/spark/blob/a131031f95dbd426516b88e1dac38965351eb501/project/plugins.sbt#L58 

There's an existing JIRA at [SPARK-14401](https://issues.apache.org/jira/browse/SPARK-14401) for removing this custom fork: I attempted this in #12310 (April 2016) but was blocked by the issue described in https://github.com/sbt/sbt-pom-reader/pull/14.

File: pom.xml
Comment (dongjoon-hyun) [inline pom.xml]: Yep. +1 for piggybacking this. I also agree that this is required to unblock https://github.com/apache/spark/pull/27279 .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.
Since @srowen sent out the email today, let's wait for two or three days for feedbacks.

**Comment by HyukjinKwon** [comment]
Thank you all. I will wait a couple of days and merge it.

File: project/SparkBuild.scala
Comment (HyukjinKwon) [inline project/SparkBuild.scala]: I added the repo to the SBT side as well to match (and fixed some comments).

**Comment by HyukjinKwon** [comment]
Let me just merge it in. I think 5 +1s from PMCs are good enough.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon and all!

**Comment by HyukjinKwon** [comment]
It has a conflict against branch-2.4. I will open a backport.

**Pull request closed without merging**
